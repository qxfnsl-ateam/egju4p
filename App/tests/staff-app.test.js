import React from 'react';
import {shallow, mount} from 'enzyme';
import renderer from 'react-test-renderer';
import StaffAppCls from '../app-src/js/staff-app-cls';
import Menu from '../app-src/js/staff-app-cls';
import DepartmentForm from '../app-src/js/department-form';
import EmployeeForm from '../app-src/js/employee-form';



describe('<StaffAppCls />', () => {
	it('Render StaffAppCls', () => {
		const departments = [{id:1, name:'Department 1'}, {id:2, name:'Department 2'}, {id:3, name:'Department 3'}];
		const employees = [{id:1, firstName:'First 1', lastName:'Last 1', departmentId:2}, {id:2, firstName:'First 2', lastName:'Last 2', departmentId:2}];
		const selectedKind = 'departments';//'employees'
		const selectedInf = {id:1};
		const expanded = 2;
		const expand = jest.fn();
		const loadStaff = jest.fn();
		const selectStaff = jest.fn();
		const newStaff = jest.fn();
		const saveStaff = jest.fn();
		const deleteStaff = jest.fn();
		const store = {subscribe:jest.fn(), dispatch:jest.fn(), getState:jest.fn()}
		const cmp = shallow(
			<StaffAppCls departments={departments} employees={employees} selectedKind={selectedKind} selectedInf={selectedInf} expanded={expanded} expand={expand} loadStaff={loadStaff} selectStaff={selectStaff} newStaff={newStaff} saveStaff={saveStaff} deleteStaff={deleteStaff} store={store} />
		);

		const eApp = cmp.find('.staff-app');
		expect(eApp.length).toBe(1);

		const eLeft = eApp.find('.staff-app-lf');
		expect(eLeft.length).toBe(1);
		const eMenu = eLeft.find('Menu');
		expect(eMenu.length).toBe(expanded ? 2 : 1);

		expect(eMenu.at(0).prop('list')).toMatchObject(departments.map(item => {
			return {value:item.id, title:item.name};
		}));
		expect(eMenu.at(0).prop('current')).toBe(selectedKind == 'departments' && selectedInf ? selectedInf.id : null);
		expect(eMenu.at(0).prop('expanded')).toBe(expanded);
		expect(eMenu.at(0).prop('prefix')).toMatch('staff-dep-');
		expect(eMenu.at(0).prop('onExpand')).toBe(expand);
		expect(eMenu.at(0).prop('onAction')).toBeDefined();
		if (expanded) {
			expect(eMenu.at(1).prop('list')).toMatchObject(employees.map(item => {
				return {value:item.id, title:`${item.lastName} ${item.firstName}`};
			}));
			expect(eMenu.at(1).prop('current')).toBe(selectedKind == 'employees' && selectedInf ? selectedInf.id : null);
			expect(eMenu.at(1).prop('expanded')).toBeFalsy();
			expect(eMenu.at(1).prop('prefix')).toMatch('staff-emp-');
			expect(eMenu.at(1).prop('onExpand')).toBeFalsy();
			expect(eMenu.at(0).prop('onAction')).toBeDefined();
		}

		const eRight = eApp.find('.staff-app-rt');
		expect(eRight.length).toBe(1);

		const eButs = eRight.find('button.staff-app-new');
		expect(eButs.length).toBe(2);
		eButs.forEach((eBut, i) => {
			expect(eBut.prop('onClick')).toBeDefined();
			eBut.simulate('click');
			expect(newStaff).toBeCalled();
		});

		const eDepForm = eRight.find('DepartmentForm');
		expect(eDepForm.length).toBe(selectedKind == 'departments' ? 1 : 0);
		if (selectedKind == 'departments') {
			expect(eDepForm.prop('entry')).toBe(selectedInf ? departments.find(item => item.id == selectedInf.id) || selectedInf : null);
			expect(eDepForm.prop('onUpdate')).toBeDefined();
			expect(eDepForm.prop('onDelete')).toBeDefined();
			expect(eDepForm.key()).toMatch(selectedInf ? '' + selectedInf.id : null);
		}

		const eEmpForm = eRight.find('EmployeeForm');
		expect(eEmpForm.length).toBe(selectedKind == 'employees' ? 1 : 0);
		if (selectedKind == 'employees') {
			expect(eEmpForm.prop('entry')).toBe(selectedInf ? employees.find(item => item.id == selectedInf.id) || selectedInf : null);
			expect(eEmpForm.prop('onUpdate')).toBeDefined();
			expect(eEmpForm.prop('onDelete')).toBeDefined();
			expect(eEmpForm.key()).toMatch(selectedInf ? '' + selectedInf.id : null);
		}

		const eNoForm = eRight.find('.staff-no-form');
		expect(eNoForm.length).toBe(!selectedKind ? 1 : 0);
	})


	it('Render StaffAppCls', () => {
		const departments = [{id:1, name:'Department 1'}, {id:2, name:'Department 2'}, {id:3, name:'Department 3'}];
		const employees = [{id:1, firstName:'First 1', lastName:'Last 1', departmentId:2}, {id:2, firstName:'First 2', lastName:'Last 2', departmentId:2}];
		const selectedKind = 'departments';//'employees'
		const selectedInf = {id:1};
		const expanded = 2;
		const expand = jest.fn();
		const loadStaff = jest.fn();
		const selectStaff = jest.fn();
		const newStaff = jest.fn();
		const saveStaff = jest.fn();
		const deleteStaff = jest.fn();
		const store = {subscribe:jest.fn(), dispatch:jest.fn(), getState:jest.fn()}
		const cmp = mount(
			<StaffAppCls departments={departments} employees={employees} selectedKind={selectedKind} selectedInf={selectedInf} expanded={expanded} expand={expand} loadStaff={loadStaff} selectStaff={selectStaff} newStaff={newStaff} saveStaff={saveStaff} deleteStaff={deleteStaff} store={store} />
		);

		expect(loadStaff).toBeCalled();
	})


	it('DOM tree', () => {
		const departments = [{id:1, name:'Department 1'}, {id:2, name:'Department 2'}, {id:3, name:'Department 3'}];
		const employees = [{id:1, firstName:'First 1', lastName:'Last 1', departmentId:2}, {id:2, firstName:'First 2', lastName:'Last 2', departmentId:2}];
		const selectedKind = 'departments';//'employees'
		const selectedInf = {id:1};
		const expanded = 2;
		const expand = jest.fn();
		const loadStaff = jest.fn();
		const selectStaff = jest.fn();
		const newStaff = jest.fn();
		const saveStaff = jest.fn();
		const deleteStaff = jest.fn();
		const store = {subscribe:jest.fn(), dispatch:jest.fn(), getState:jest.fn()}
		const cmp = renderer.create(
			<StaffAppCls departments={departments} employees={employees} selectedKind={selectedKind} selectedInf={selectedInf} expanded={expanded} expand={expand} loadStaff={loadStaff} selectStaff={selectStaff} newStaff={newStaff} saveStaff={saveStaff} deleteStaff={deleteStaff} store={store} />
		);

		expect(cmp.toJSON()).toMatchSnapshot();
	});
});
