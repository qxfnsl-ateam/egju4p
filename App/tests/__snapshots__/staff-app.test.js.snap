// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StaffAppCls /> DOM tree 1`] = `
<div
  className="staff-app container"
>
  <aside
    className="staff-app-lf col-sm-4"
  >
    <div
      className="panel panel-primary"
    >
      <div
        className="panel panel-heading"
      >
        <h4
          className="panel-title"
        >
          Отделы и сотрудники
        </h4>
      </div>
      <ul
        className="staff-menu list-group"
      >
        <li
          className="staff-menu-item on list-group-item"
          onClick={[Function]}
          value={1}
        >
          Department 1
          <button
            className="staff-menu-exp"
            onClick={[Function]}
          />
        </li>
        <li
          className="staff-menu-item list-group-item"
          onClick={[Function]}
          value={2}
        >
          Department 2
          <button
            className="staff-menu-exp on"
            onClick={[Function]}
          />
          <ul
            className="staff-menu list-group"
          >
            <li
              className="staff-menu-item list-group-item"
              onClick={[Function]}
              value={1}
            >
              Last 1 First 1
            </li>
            <li
              className="staff-menu-item list-group-item"
              onClick={[Function]}
              value={2}
            >
              Last 2 First 2
            </li>
          </ul>
        </li>
        <li
          className="staff-menu-item list-group-item"
          onClick={[Function]}
          value={3}
        >
          Department 3
          <button
            className="staff-menu-exp"
            onClick={[Function]}
          />
        </li>
      </ul>
    </div>
  </aside>
  <div
    className="staff-app-rt col-sm-8"
  >
    <h2
      className="staff-app-title text-primary"
    >
      Персонал
    </h2>
    <button
      className="staff-app-new staff-app-new-dep btn btn-default"
      onClick={[Function]}
    >
      Новый отдел
    </button>
    <button
      className="staff-app-new staff-app-new-emp btn btn-default"
      onClick={[Function]}
    >
      Новый сотрудник
    </button>
    <div
      className="panel panel-default"
    >
      <div
        className="panel panel-body"
      >
        <form
          className="staff-form staff-form-dep"
        >
          <h4
            className="staff-form-title text-primary"
          >
            Изменить информацию об отделе
          </h4>
          <input
            name="id"
            type="hidden"
            value={1}
          />
          <div
            className="form-group"
          >
            <label
              className="staff-dep-name"
              htmlFor="staff-dep-name"
            >
              Название
            </label>
            <input
              className="form-control"
              defaultValue="Department 1"
              id="staff-dep-name"
              name="name"
              pattern="[^\\\\\\\\s]{3,}"
              required={true}
              type="text"
            />
          </div>
          <button
            className="staff-form-update btn btn-default"
            onClick={[Function]}
          >
            Сохранить
          </button>
          <button
            className="staff-form-delete btn btn-danger"
            onClick={[Function]}
          >
            Удалить
          </button>
          <button
            className="staff-form-reset btn btn-default"
            type="reset"
          >
            Сбросить
          </button>
        </form>
      </div>
    </div>
  </div>
</div>
`;
